### PR TITLE
OnOff: add animations and drag support

### PR DESCRIPTION
**Pull request by TomaszLukawskiSam**
[Issue] https://github.com/Samsung/TAU/issues/889
[Problem] Controllers -> On/Off switch animations
[Solution]
 - added elastic animation for on-off switch

[Demo]
onoff

Signed-off-by: Tomasz Lukawski <t.lukawski@samsung.com>

### CODE REVIEW

**Comment by TomaszLukawskiSam** [comment]
@lslachciak 
Hi, could you check this change? Thanks

File: src/js/core/widget/core/OnOffSwitch.js [hunk 1]
```diff
@@ -101,8 +101,13 @@
...
 					self.options = {
 						appearance: "slider"
-					},
+					};
 					self._ui = {};
+					self._callbacks = {};
+					self._transform = {
+						bgColor: {from: {r: 0, g: 0, b: 0, a: 0}, to: {r: 3, g: 129, b: 254, a: 1}},
+						borderColor: {from: {r: 143, g: 143, b: 143, a: 1}, to: {r: 0, g: 0, b: 0, a: 0}}
+					}
```
Comment (HunseopJeong) [inline src/js/core/widget/core/OnOffSwitch.js]: Need a semicolon

File: src/js/core/widget/core/OnOffSwitch.js [hunk 2]
```diff
@@ -386,6 +399,136 @@
...
+			OnOffSwitch.prototype._onDragStart = function () {
+				var self = this,
+					ui = self._ui,
+					checkbox = ui.input,
+					handler = ui.handler,
+					inputRect = checkbox.getBoundingClientRect(),
+					handleRect = handler.getBoundingClientRect(),
+					inputStyle = window.getComputedStyle(checkbox),
+					border = parseInt(inputStyle.borderLeftWidth, 10) +
+						parseInt(inputStyle.borderLeftWidth, 10);
```
Comment (HunseopJeong) [inline src/js/core/widget/core/OnOffSwitch.js]: Is it correct to add the borderLeftwidth twice?

File: src/js/core/widget/core/OnOffSwitch.js [hunk 3]
```diff
@@ -398,16 +541,44 @@
...
+				input.addEventListener("keyup", self, true);
+				input.addEventListener("dragstart", self, true);
+				input.addEventListener("drag", self, true);
+				input.addEventListener("dragend", self, true);
+				events.on(self._ui.handler, "animationend animationEnd webkitAnimationEnd", self, false);
+			};
+
+			/**
+			 * Binds events to widget
+			 * @method _bindEvents
```
Comment (HunseopJeong) [inline src/js/core/widget/core/OnOffSwitch.js]: Rename with '_unbindEvents'

File: src/js/core/widget/core/OnOffSwitch.js [hunk 4]
```diff
@@ -398,16 +541,44 @@
...
+			 * @protected
+			 * @member ns.widget.core.OnOffSwitch
+			 * @instance
```
Comment (HunseopJeong) [inline src/js/core/widget/core/OnOffSwitch.js]: Is it correct to define the @instance?

File: src/js/core/widget/core/OnOffSwitch.js [hunk 5]
```diff
@@ -386,6 +399,136 @@
...
+					ui = self._ui,
+					checkbox = ui.input,
+					handler = ui.handler,
+					inputRect = checkbox.getBoundingClientRect(),
+					handleRect = handler.getBoundingClientRect(),
+					inputStyle = window.getComputedStyle(checkbox),
+					border = parseInt(inputStyle.borderLeftWidth, 10) +
+						parseInt(inputStyle.borderLeftWidth, 10);
+
+				self._moveWidth = inputRect.width - handleRect.width + border;
```
Comment (DohyungLim) [inline src/js/core/widget/core/OnOffSwitch.js]: If you use border only in this function, you may want to fill it out immediately.

File: src/js/core/widget/core/OnOffSwitch.js [hunk 6]
```diff
@@ -386,6 +399,136 @@
...
+					case "drag":
+						self._onDrag(event);
+						break;
+					case "dragstart":
+						self._onDragStart(event);
+						break;
+					case "dragend":
+						self._onDragEnd(event);
+						break;
+					case "animationend":
```
Comment (DohyungLim) [inline src/js/core/widget/core/OnOffSwitch.js:527]: Do you use both "end" and "End"?

**Review by DohyungLim — COMMENTED** [review]
2 comment.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 2 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: You are right. Thanks.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 4 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: I'll remove it.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 6 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js:527]: Yes, it is intended. It happened that some versions of the web engine on  Tizen generated the event with this name.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 5 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: Ok. I'll refactor this.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 1 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: Done.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 2 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: Done.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 4 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: Done.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 5 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: Done.

File: src/js/core/widget/core/OnOffSwitch.js (hunk 3 above)
Comment (TomaszLukawskiSam) [reply src/js/core/widget/core/OnOffSwitch.js]: Done.

**Review by lslachciak — APPROVED** [review]
works fine, seems ok

**Pull request merged**